text/0615-autotracking-memoization.md

```diff
@@ -0,0 +1,385 @@
+- Start Date: 2020-04-17
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/615
+- Tracking: (leave this empty)
+
+# Autotracking Memoization
+
+## Summary
+
+Provides a low-level primitive for memoizing the result of a function based on
+autotracking, allowing users to create their own reactive systems that can
+respond to changes in autotracked state
+
+```js
+import { tracked } from '@glimmer/tracking';
+import { createCache, getValue } from '@glimmer/tracking/primitives/cache';
+
+let count = 0;
+
+class Person {
+  @tracked firstName = 'Jen';
+  @tracked lastName = 'Weber';
+
+  #fullName = createCache(() => {
+    ++count;
+    return `${this.firstName} ${this.lastName}`;
+  })
+
+  get fullName() {
+    return getValue(this.#fullName);
+  }
+}
+
+let person = new Person();
+
+console.log(person.fullName); // Jen Weber
+console.log(count); // 1;
+console.log(person.fullName); // Jen Weber
+console.log(count); // 1;
+
+person.firstName = 'Jennifer';
+
+console.log(person.fullName); // Jennifer Weber
+console.log(count); // 2;
+```
+
+## Motivation
+
+Autotracking is the fundamental reactivity model within Ember Octane, and has
+been highly successful so far in its usage and adoption. However, users today
+can only integrate with autotracking via the `@tracked` decorator, which allows
+them to _create_ tracked root state. There is no way to write code that responds
+to changes in that root state directly - the only way to do so is indirectly via
+Ember's templating layer.
+
+An example of where users might want to do this is the [`@cached`](https://github.com/emberjs/rfcs/pull/566)
+decorator for getters. This decorator only reruns its code when the tracked
+state it accessed during its previous computation changes. Currently, it would
+not be possible to build this decorator with public APIs.
```

Comment:
> Currently, it would not be possible to build this decorator with public APIs.

Is this actually true? Couldn't you use a combination of `computed` + `dependentKeyCompat`? I agree that it doesn't _really_ matter, but it would be good to avoid the absolutist terminology if it is actually not correct.

Maybe:

```suggestion
state it accessed during its previous computation changes. Currently, it would
be quite difficult and error prone to build this decorator with public APIs.
```